perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -904,6 +931,59 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
 	public void addMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationExistsException {
 		checkParentVos(sess, vo, memberVo);
 		vosManagerImpl.addMemberVo(sess, vo, memberVo);
+
+		List<Member> memberVoMembers = perunBl.getMembersManagerBl().getMembers(sess, memberVo);
+		AttributeDefinition memberExpirationAttrDef;
+
+		try {
+			memberExpirationAttrDef = perunBl.getAttributesManagerBl().getAttributeDefinition(sess, AttributesManager.NS_MEMBER_ATTR_DEF + ":membershipExpiration");
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		for (Member member : memberVoMembers) {
+			if (member.getStatus() == Status.DISABLED || member.getStatus() == Status.INVALID) {
+				continue;
+			}
+			try {
+
+				try {
+					Member existingMember = perunBl.getMembersManagerBl().getMemberByUserId(sess, vo, member.getUserId());
+					if (member.getStatus() == Status.EXPIRED) {
+						// member is expired in memberVo but exists in parentVo, don't update
+						continue;
+					}
+
+					// member is valid in memberVo, so reset expiration in parentVo
+					perunBl.getAttributesManagerBl().removeAttribute(sess, existingMember, memberExpirationAttrDef);
+					if (existingMember.getStatus() != Status.VALID) {
+						perunBl.getMembersManagerBl().validateMemberAsync(sess, existingMember);
+					}
+					//update memberOrganizations
+					Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, existingMember, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+					ArrayList<String> currentValue = attribute.valueAsList();
+					currentValue.add(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, existingMember, attribute);
+
+				} catch (MemberNotExistsException e) {
+					// if user is member only in member vo, create member in parent vo
+					Member newMember = perunBl.getMembersManagerBl().createMember(sess, vo, perunBl.getUsersManagerBl().getUserByMember(sess, member));
+					if (member.getStatus() == Status.VALID) {
+						// remove expiration set according to parentVo and update memberOrganizations
+						perunBl.getAttributesManagerBl().removeAttribute(sess, newMember, memberExpirationAttrDef);
+						Attribute attribute = new Attribute(perunBl.getAttributesManagerBl().getAttributeDefinition(sess, A_MEMBER_DEF_MEMBER_ORGANIZATIONS));
+						ArrayList<String> newValue = new ArrayList<>(List.of(memberVo.getShortName()));
+						attribute.setValue(newValue);
+						perunBl.getAttributesManagerBl().setAttribute(sess, newMember, attribute);
+					}
+					perunBl.getMembersManagerBl().validateMemberAsync(sess, newMember);
+				}
+
+			} catch (WrongAttributeAssignmentException | WrongReferenceAttributeValueException | AttributeNotExistsException | WrongAttributeValueException | AlreadyMemberException | ExtendMembershipException e) {
```

Comment:
I am thinking about `ExtendMembershipException` - maybe we could silently skip these members (e.g. this exception is thrown when they don't have required loa) instead of the whole operation of adding a member organization to fail. Or perhaps creating such members in a different way because they shouldn't follow the parent VO's expiration lifecycle (and this exception is thrown because of the parent VO's rules). Just a thought.